src/notify/notify.c

```diff
@@ -310,56 +336,254 @@ static int parse_argv(int argc, char *argv[]) {
+                ret = log_debug_errno(SYNTHETIC_ERRNO(EPROTO),
+                                      "Child terminated by signal %s.", signal_to_string(si->si_status));
+        else
+                ret = log_debug_errno(SYNTHETIC_ERRNO(EPROTO),
+                                      "Child terminated due to unknown reason.");
+
+        r = sd_event_exit(sd_event_source_get_event(s), ret);
+        if (r < 0)
+                return log_error_errno(r, "Failed to request event loop exit: %m");
+
+        return 0;
+}
+
+static int action_fork(char *const *_command) {
+
+        static const int forward_signals[] = {
+                SIGHUP,
+                SIGTERM,
+                SIGINT,
+                SIGQUIT,
+                SIGTSTP,
+                SIGCONT,
+                SIGUSR1,
+                SIGUSR2,
+        };
+
+        int r;
+
+        assert(!strv_isempty(_command));
+
+        /* Make a copy, since pidref_safe_fork_full() will change argv[] further down. */
+        _cleanup_strv_free_ char **command = strv_copy(_command);
+        if (!command)
+                return log_oom();
+
+        _cleanup_close_ int socket_fd = socket(AF_UNIX, SOCK_DGRAM|SOCK_CLOEXEC|SOCK_NONBLOCK, 0);
+        if (socket_fd < 0)
+                return log_error_errno(errno, "Failed to allocate AF_UNIX socket for notifications: %m");
+
+        r = setsockopt_int(socket_fd, SOL_SOCKET, SO_PASSCRED, true);
+        if (r < 0)
+                return log_error_errno(r, "Failed to enable SO_PASSCRED: %m");
+
+        r = setsockopt_int(socket_fd, SOL_SOCKET, SO_PASSPIDFD, true);
+        if (r < 0)
+                log_debug_errno(r, "Failed to enable SO_PASSPIDFD, ignoring: %m");
+
+        /* Pick an address via auto-bind */
+        union sockaddr_union sa = {
+                .sa.sa_family = AF_UNIX,
+        };
+        if (bind(socket_fd, &sa.sa, offsetof(union sockaddr_union, un.sun_path)) < 0)
+                return log_error_errno(errno, "Failed to bind AF_UNIX socket: %m");
+
+        _cleanup_free_ char *addr_string = NULL;
+        r = getsockname_pretty(socket_fd, &addr_string);
+        if (r < 0)
+                return log_error_errno(r, "Failed to get socket name: %m");
+
+        _cleanup_free_ char *c = strv_join(command, " ");
+        if (!c)
+                return log_oom();
+
+        _cleanup_(pidref_done) PidRef child = PIDREF_NULL;
+        r = pidref_safe_fork_full(
+                        "(notify)",
+                        /* stdio_fds= */ (const int[]) { -EBADF, -EBADF, STDERR_FILENO },
+                        /* execpt_fds= */ NULL,
+                        /* n_except_fds= */ 0,
+                        /* flags= */ FORK_REARRANGE_STDIO,
+                        &child);
+        if (r < 0)
+                return log_error_errno(r, "Failed to fork child in order to execute '%s': %m", c);
+        if (r == 0) {
+                socket_fd = safe_close(socket_fd);
+                pidref_done(&child);
```

Comment:
Is this necessary with CLOEXEC?

Reply:
no, it's not. But i really like to keep them in, so that once can validate that everything is in order when debugging by looking at /proc/self/fd/.

i'll add a comment.